lib/internal/modules/run_main.js

```diff
@@ -9,14 +9,19 @@ const path = require('path');
 
 /**
  * Get the absolute path to the main entry point.
- * @param {string} main Entry point path
+ * @param {string} main - Entry point path
  */
 function resolveMainPath(main) {
-  // Note extension resolution for the main entry point can be deprecated in a
-  // future major.
-  // Module._findPath is monkey-patchable here.
-  const { Module } = require('internal/modules/cjs/loader');
-  let mainPath = Module._findPath(path.resolve(main), null, true);
+  /** @type {string} */
+  let mainPath;
+  if (getOptionValue('--experimental-default-type') === 'module') {
+    mainPath = path.resolve(main);
```

Comment:
This should not be needed, `pathToFileURL` would resolve the path.
```suggestion
    return;
```

Reply:
It feels wrong to have a function called `resolveMainPath` that sometimes just doesn’t bother to resolve the main path. I can understand returning a falsy value if the main path _can’t_ be resolved, but if it can, we should return it.

Alternatively we could skip calling `resolveMainPath` at all for the new flow, if it’s not needed.

Reply:
Maybe rename it to `resolveMainPathIfApplicable`

Reply:
I updated to just not use `resolveMainPath` or `shouldUseESMLoader` in the new flow, where they’re not needed.

Reply:
Can you remove this now please?

Reply:
I guess we want to take `--preserve-symlinks-main` into account, so it would be
```suggestion
    if (getOptionValue('--preserve-symlinks-main')) return;
    mainPath = path.resolve(main);
```

Reply:
It's an optimisation: make 1 `path.resolve` call instead of 2 (the other one being done by `pathToFileURL`)

Reply:
Okay I think I got it, please take a look at the latest commit.